packages/aws-event-stream/src/aws_event_stream/_private/deserializers.py

```diff
@@ -40,9 +40,18 @@ def __init__(
         self._source = source
         self._is_client_mode = is_client_mode
         self._deserializer = deserializer
+        self._closed = False
 
     async def receive(self) -> E | None:
-        event = await Event.decode_async(self._source)
+        if self._closed:
+            return None
+
+        try:
+            event = await Event.decode_async(self._source)
+        except Exception as e:
+            await self.close()
+            raise IOError("Failed to read from stream.") from e
```

Comment:
Exceptions could be raised here for other decode issues right (ie: like an invalid message), rather than just failing to read from the stream?